File: leetcode/go/740.delete-and-earn.go
func deleteAndEarn(nums []int) int {
    m := make(map[int]int)
    max_num := 0
    for _, n := range nums {
        m[n] += 1
        if n > max_num {
            max_num = n
        }
    }
    dp := make([]int, max_num+1)
    if val, ok := m[1]; ok {
        dp[1] = val
    }
    for i := 2; i < max_num+1; i++ {
        if val, ok := m[i]; ok {
            dp[i] = Max(dp[i-1], (dp[i-2] + i * val))
        } else {
            dp[i] = dp[i-1]
        }
    }
    return dp[max_num]
}

func Max(a int, b int) int {
    if a > b {return a}
    return b
}

File: leetcode/rust-leetcode/src/740.delete-and-earn.rs
impl Solution {
    pub fn delete_and_earn(nums: Vec<i32>) -> i32 {
        let mx = *nums.iter().max().unwrap() as usize;
        let mut counts = vec![0; mx+1];
        for &n in nums.iter() {
            counts[n as usize] += 1;
        }
        let mut scores = vec![0; mx + 1];
        scores[1] = counts[1];
        for i in 2..=mx {
            scores[i] = scores[i-1].max(scores[i-2] + counts[i] * i);
        }
        scores[mx] as _
    }
}

